### PR TITLE
fix: StringUtils: bugs in edge cases and enhance unit test cases

### PR DESCRIPTION
**Pull request by miurahr**
When extending unit test to improve test coverage, there are several bugs in edge cases that should be found in unit test level.
The topic branch also refactoring StringUtils class methods.

## Pull request type

- Bug fix -> [bug]

## Which ticket is resolved?

There is no issue to point

## What does this PR change?

- Refactoring for Clarity and Performance:
    - Replaced for-loops for iterating over code points with while-loops.
    - Reorganized and clarified switch cases for better maintainability and readability.
    - Improved comments and JavaDocs for the utility functions.

- Expanded Functionality:
    - New helper methods for managing XML characters and handling various string operations.
    - Additional utility constants introduced for clarity and reusability.

- Null Safety Enhancements:
    - Added annotations like @NotNull to enforce safer parameter handling.
    - Modified StringUtilTest to ensure NullPointerException is tested in relevant cases.

- Tests Update and Coverage:
    - Revised and added comprehensive unit tests in StringUtilTest to cover new functions, edge cases, and locale-specific behaviors.
    - Split large tests into smaller, more manageable ones for enhanced clarity.

### CODE REVIEW

**Comment by miurahr** [comment]
The PR should be split to 3 parts:  fix the bug and corresponding unit test, refactoring the code, and updating of the code styles and javadoc.

**Comment by miurahr** [comment]
StringUtil is too big, especially for the `normalizeWidth` method. It is better to create a properties file for conversion data and import it into the class.

File: test/src/org/omegat/util/StringUtilTest.java [hunk 1]
```diff
@@ -428,4 +556,121 @@ public void testCompareToNullable() {
         assertEquals(32, StringUtil.compareToNullable("a", "A"));
         assertEquals(-32, StringUtil.compareToNullable("A", "a"));
     }
+
+    @Test
+    public void testReplaceSquaredLatinAbbreviations() {
```
Comment (miurahr) [inline test/src/org/omegat/util/StringUtilTest.java:561]: It is better to write functional test for `normalizeWidth` method for future replacement of the functionality with 3rd party library.
